cilium: Rework neighbor handling (#17713)
cilium-team/cilium: Documentation/gettingstarted/kubeproxy-free.rst

Neighbor Discovery
******************

When kube-proxy replacement is enabled, Cilium does L2 neighbor discovery of nodes

> in the cluster. This is required for the service load-balancing in order to populate
> L2 addresses for backends for the fast-path where it is not possible to dynamically
> resolve neighbors on demand.

Reviewer: ```suggestion
in the cluster. This is required for the service load-balancing to populate L2
addresses for backends since it is not possible to dynamically resolve neighbors
on demand in the fast-path.
```

In Cilium 1.10 or earlier, the agent itself contained an ARP resolution library
where it triggered discovery and periodic refresh of new nodes joining the cluster.
The resolved neighbor entries were pushed into the kernel and refreshed as PERMANENT
entries. In some rare cases, Cilium 1.10 or earlier might have left stale entries behind

> in the neighbor table causing packets between some nodes to be dropped. To prevent from
> performing the neighbor discovery and instead relying on the Linux kernel to discover
> hosts on the same L2 network you can pass the ``--enable-l2-neigh-discovery=false``

Reviewer: ```suggestion
in the neighbor table causing packets between some nodes to be dropped. To skip the
neighbor discovery and instead rely on the Linux kernel to discover neighbors, you can
pass the ``--enable-l2-neigh-discovery=false``
```

flag to the cilium-agent. However, note that relying on the Linux Kernel might also

> cause some packets to be dropped, for example, a NodePort request can be dropped on

Reviewer: You can't join two independent clauses with a just comma:
```suggestion
cause some packets to be dropped. For example, a NodePort request can be dropped on
```
(Or use a semi-colon.)

an intermediate node (i.e., the one which received a service packet and is going to
forward it to a destination node which runs the selected service endpoint). This
could happen if there is no L2 neighbor entry in the kernel (due to the entry being
garbage collected or given that the neighbor resolution has not been done by the kernel).
This is because it is not possible to drive the neighbor resolution from BPF programs

> in the fast-path e.g. at the XDP layer.

Reviewer: Overall, this new documentation section feels like exactly what we'd put in the release blog post. It's a bit unusual I think to have the documentation for a given version (v1.11) describe the behavior of previous versions.

Author: I think it may be useful given most folks are still on v1.10 or earlier and these are two different approaches, so might be good to know what changed in terms of neighbor discovery. We can remove the before/after in 1.12, but I think it'd be good to have this info here.

From Cilium 1.11 onwards, the neighbor discovery has been fully reworked and the Cilium
internal ARP resolution library has been removed from the agent. The agent now fully
relies on the Linux kernel to discover gateways or hosts on the same L2 network. As per
our recent kernel work `presented at Plumbers <https://linuxplumbersconf.org/event/11/contributions/953/>`__,
"managed" neighbor entries have been `upstreamed <https://lore.kernel.org/netdev/20211011121238.25542-1-daniel@iogearbox.net/>`__
and will be available in Linux kernel v5.16 or later which the Cilium agent will detect
and transparently use. In this case, the agent pushes down L3 addresses of new nodes
joining the cluster as externally learned "managed" neighbor entries. For introspection,

> iproute2 displays them as "managed extern_learn". The "extern_learn" attribute is present
> such that the entries are not garbage collected by the kernel's neighboring subsystem

Reviewer: ```suggestion
iproute2 displays them as "managed extern_learn". The "extern_learn" attribute prevents
garbage collection of the entries by the kernel's neighboring subsystem. Such "managed"
```

garbage collector. Such "managed" neighbor entries are dynamically resolved and periodically
refreshed by the Linux kernel itself in case there is no active traffic for a certain period

> of time, meaning, the kernel attempts to always keep them in REACHABLE state. For Linux

Reviewer: Ditto, two independent clauses joined by a comma:
```suggestion
of time. That is, the kernel attempts to always keep them in REACHABLE state. For Linux
```

kernels v5.15 or earlier where "managed" neighbor entries are not present, the Cilium
agent similarly pushes L3 addresses of new nodes into the kernel for dynamic resolution,
but with an agent triggered periodic refresh. For introspection, iproute2 displays them
only as "extern_learn" in this case. If there is no active traffic for a certain period
of time, then a Cilium agent controller triggers the Linux kernel-based re-resolution for
attempting to keep them in REACHABLE state. The refresh interval can be changed if needed
through a ``--arping-refresh-period=30s`` flag passed to the cilium-agent. The default

> period by the agent is ``30s`` which corresponds to the kernel's base reachable time.

Reviewer: ```suggestion
period is ``30s`` which corresponds to the kernel's base reachable time.
```
